pilot/pkg/model/config_test.go

```diff
@@ -177,42 +178,70 @@ func TestServiceKey(t *testing.T) {
 func TestSubsetKey(t *testing.T) {
 	hostname := model.Hostname("hostname")
 	cases := []struct {
+		got      string
 		hostname model.Hostname
 		subset   string
 		port     *model.Port
 		want     string
+		err      error
 	}{
 		{
 			hostname: "hostname",
 			subset:   "subset",
 			port:     &model.Port{Name: "http", Port: 80, Protocol: model.ProtocolHTTP},
-			want:     "outbound|http|subset|hostname",
+			want:     "outbound.http.subset.hostname",
 		},
 		{
 			hostname: "hostname",
 			subset:   "subset",
 			port:     &model.Port{Port: 80, Protocol: model.ProtocolHTTP},
-			want:     "outbound||subset|hostname",
+			want:     "outbound.80.subset.hostname",
 		},
 		{
 			hostname: "hostname",
 			subset:   "",
 			port:     &model.Port{Name: "http", Port: 80, Protocol: model.ProtocolHTTP},
-			want:     "outbound|http||hostname",
+			want:     "outbound.http._default_.hostname",
+		},
+		{
+			want:   "badClusterKey",
+			subset: "",
+			port:   &model.Port{Name: "dontcare", Port: 80, Protocol: model.ProtocolHTTP},
+			got:    "badClusterKey",
+			err:    fmt.Errorf("malformed cluster-key"),
 		},
 	}
 
 	for _, c := range cases {
-		got := model.BuildSubsetKey(model.TrafficDirectionOutbound, c.subset, hostname, c.port)
-		if got != c.want {
-			t.Errorf("Failed: got %q want %q", got, c.want)
-		}
+		t.Run(c.want, func(t *testing.T) {
+			got := c.got
+			if got == "" {
+				got = model.BuildSubsetKey(model.TrafficDirectionOutbound, c.subset, hostname, c.port)
+				if got != c.want {
+					t.Errorf("Failed: got %q want %q", got, c.want)
+				}
+			}
+			// test parse subset key. ParseSubsetKey is the inverse of BuildSubsetKey
+			_, s, h, p, err := model.ParseSubsetKey(got)
 
-		// test parse subset key. ParseSubsetKey is the inverse of BuildSubsetKey
-		_, s, h, p := model.ParseSubsetKey(got)
-		if s != c.subset || h != c.hostname || p.Name != c.port.Name {
-			t.Errorf("Failed: got %s,%s,%s want %s,%s,%s", s, h, p.Name, c.subset, c.hostname, c.port.Name)
-		}
+			wantErr := c.err != nil
+			gotErr := err != nil
+
+			if gotErr != wantErr {
```

Comment:
tip: boilerplate pattern for this (makes for easy copying to other tests):
if got, want := c.err != nil, err != nil; got != want {
  t.Fatalf("\n got: %v\nwant: %v", got, want)
}
what you have is fine but consider for future tests.

Reply:
I get a bit dizzy trying to read the shorter version - I would avoid perl in go.